File: agent/agent-harness/agent-harness/frontend/src/api/diagnosis.ts
import { api } from "./client";

/**
 * Bottleneck Diagnosis Loop adapter (harness/api/diagnosis.py, doc03 §9).
 * Unlike engineering-design, most *content* here (hypotheses, tests,
 * decisions) is produced by `harness/orchestrator/adapters.py` driving the
 * service layer directly - the HTTP surface only exposes read access to
 * that content plus the narrow set of human-in-the-loop writes the router
 * actually defines (create session, link evidence, run a model, approve a
 * decision, and the pure state-transition `action` endpoint). Do not add
 * "generate hypotheses"/"create test" calls here - no such route exists.
 */

export interface DiagnosisSessionSummary {
  diagnosisSessionId: string;
  projectId: string;
  status: string;
  dataSufficiency: string;
  approvalState: string;
  biologicalSystem: Record<string, unknown>;
  createdAt: number;
  updatedAt: number;
}

export interface DiagnosisSessionDetail {
  diagnosisSessionId: string;
  projectId: string;
  status: string;
  dataSufficiency: string;
  approvalState: string;
  activeHypothesisSetVersion: number;
  biologicalSystem: Record<string, unknown>;
  baselineObservationIds: string[];
  version: number;
}

export interface HypothesisRow {
  hypothesisVersionId: string;
  status: string;
  mechanismClass: string | null;
  statement: string | null;
  explanatoryCoverage: Record<string, unknown>;
  contradictions: unknown[];
}

export interface EvidenceLinkRow {
  evidenceLinkId: string;
  hypothesisVersionId: string;
  evidenceItemId: string;
  relation: string;
  claim: string;
}

export interface ModelCapability {
  available: boolean;
  reason: string;
}

export interface ModelRunResult {
  modelRunId: string;
  capabilityStatus: string;
  runtimeStatus: string;
  outputs: Record<string, unknown>;
  logSummary: string;
}

export interface DiagnosticTestRow {
  testId: string;
  assay: string;
  status: string;
  discriminatesHypotheses: boolean;
  expectedInformationGain: string;
  plans: Array<{ planId: string; readiness: string }>;
}

export interface DecisionRow {
  decisionId: string;
  diagnosisVersion: number;
  stoppingReason: string;
  allowedNextAction: string;
  handoffStatus: string;
  leadingHypothesisIds: string[];
  alternativesNotExcludedIds: string[];
}

export interface DiagnosisTransitionRow {
  state: string;
  selectedNextState: string | null;
  gateResult: Record<string, unknown> | null;
  startedAt: number;
}

export interface ReportSection {
  title: string;
  content: Record<string, unknown>;
  traceIds: string[];
}

function toSessionSummary(r: {
  diagnosis_session_id: string; project_id: string; status: string; data_sufficiency: string;
  approval_state: string; biological_system: Record<string, unknown>; created_at: number; updated_at: number;
}): DiagnosisSessionSummary {
  return {
    diagnosisSessionId: r.diagnosis_session_id, projectId: r.project_id, status: r.status,
    dataSufficiency: r.data_sufficiency, approvalState: r.approval_state,
    biologicalSystem: r.biological_system, createdAt: r.created_at, updatedAt: r.updated_at,
  };
}

/** `GET /api/diagnosis/sessions?project_id=` - newest first. */
export async function listSessionsForProject(projectId: string): Promise<DiagnosisSessionSummary[]> {
  const raw = await api.get<{ sessions: Parameters<typeof toSessionSummary>[0][] }>(
    `/api/diagnosis/sessions?project_id=${encodeURIComponent(projectId)}`,
  );
  return raw.sessions.map(toSessionSummary);
}

export interface CreateSessionInput {
  projectId: string;
  actorId: string;
  workflowRunId?: string;
  triggeringFailureCaseId?: string;
  objectiveId?: string;
  biologicalSystem?: Record<string, unknown>;
  baselineObservationIds?: string[];
}

/** `POST /api/diagnosis/sessions`. */
export async function createSession(input: CreateSessionInput): Promise<{ diagnosisSessionId: string; status: string }> {
  const raw = await api.post<{ diagnosis_session_id: string; status: string }>("/api/diagnosis/sessions", {
    project_id: input.projectId, actor_id: input.actorId, workflow_run_id: input.workflowRunId ?? null,
    triggering_failure_case_id: input.triggeringFailureCaseId ?? null, objective_id: input.objectiveId ?? null,
    biological_system: input.biologicalSystem ?? {}, baseline_observation_ids: input.baselineObservationIds ?? [],
  });
  return { diagnosisSessionId: raw.diagnosis_session_id, status: raw.status };
}

/** `GET /api/diagnosis/sessions/{id}`. */
export async function getSession(diagnosisSessionId: string): Promise<DiagnosisSessionDetail> {
  const r = await api.get<{
    diagnosis_session_id: string; project_id: string; status: string; data_sufficiency: string;
    approval_state: string; active_hypothesis_set_version: number; biological_system: Record<string, unknown>;
    baseline_observation_ids: string[]; version: number;
  }>(`/api/diagnosis/sessions/${diagnosisSessionId}`);
  return {
    diagnosisSessionId: r.diagnosis_session_id, projectId: r.project_id, status: r.status,
    dataSufficiency: r.data_sufficiency, approvalState: r.approval_state,
    activeHypothesisSetVersion: r.active_hypothesis_set_version, biologicalSystem: r.biological_system,
    baselineObservationIds: r.baseline_observation_ids, version: r.version,
  };
}

/** `GET /api/diagnosis/sessions/{id}/hypotheses`. */
export async function listHypotheses(diagnosisSessionId: string): Promise<HypothesisRow[]> {
  const raw = await api.get<{
    hypotheses: Array<{
      hypothesis_version_id: string; status: string; mechanism_class: string | null; statement: string | null;
      explanatory_coverage: Record<string, unknown>; contradictions: unknown[];
    }>;
  }>(`/api/diagnosis/sessions/${diagnosisSessionId}/hypotheses`);
  return raw.hypotheses.map((h) => ({
    hypothesisVersionId: h.hypothesis_version_id, status: h.status, mechanismClass: h.mechanism_class,
    statement: h.statement, explanatoryCoverage: h.explanatory_coverage, contradictions: h.contradictions,
  }));
}

/** `GET /api/diagnosis/sessions/{id}/evidence`. */
export async function listEvidence(diagnosisSessionId: string): Promise<EvidenceLinkRow[]> {
  const raw = await api.get<{
    evidence_links: Array<{ evidence_link_id: string; hypothesis_version_id: string; evidence_item_id: string; relation: string; claim: string }>;
  }>(`/api/diagnosis/sessions/${diagnosisSessionId}/evidence`);
  return raw.evidence_links.map((e) => ({
    evidenceLinkId: e.evidence_link_id, hypothesisVersionId: e.hypothesis_version_id,
    evidenceItemId: e.evidence_item_id, relation: e.relation, claim: e.claim,
  }));
}

export interface LinkEvidenceInput {
  hypothesisVersionId: string;
  evidenceItemId: string;
  relation: string;
  actorId: string;
  claim?: string;
  conditionMatch?: string;
}

/** `POST /api/diagnosis/evidence-links`. */
export async function linkEvidence(input: LinkEvidenceInput): Promise<string> {
  const raw = await api.post<{ evidence_link_id: string }>("/api/diagnosis/evidence-links", {
    hypothesis_version_id: input.hypothesisVersionId, evidence_item_id: input.evidenceItemId,
    relation: input.relation, actor_id: input.actorId, claim: input.claim ?? "",
    condition_match: input.conditionMatch ?? "unknown",
  });
  return raw.evidence_link_id;
}

/** `GET /api/diagnosis/model-capabilities`. */
export async function listModelCapabilities(): Promise<Record<string, ModelCapability>> {
  return api.get<Record<string, ModelCapability>>("/api/diagnosis/model-capabilities");
}

export interface RunModelInput {
  projectId: string;
  diagnosisSessionId: string;
  adapterName: string;
  actorId: string;
  inputs?: Record<string, unknown>;
  context?: Record<string, unknown>;
  constraintsObjectiveParameters?: Record<string, unknown>;
}

/** `POST /api/diagnosis/model-runs`. */
export async function runModel(input: RunModelInput): Promise<ModelRunResult> {
  const r = await api.post<{ model_run_id: string; capability_status: string; runtime_status: string; outputs: Record<string, unknown>; log_summary: string }>(
    "/api/diagnosis/model-runs",
    {
      project_id: input.projectId, diagnosis_session_id: input.diagnosisSessionId, adapter_name: input.adapterName,
      actor_id: input.actorId, inputs: input.inputs ?? {}, context: input.context ?? {},
      constraints_objective_parameters: input.constraintsObjectiveParameters ?? {},
    },
  );
  return { modelRunId: r.model_run_id, capabilityStatus: r.capability_status, runtimeStatus: r.runtime_status, outputs: r.outputs, logSummary: r.log_summary };
}

/** `GET /api/diagnosis/sessions/{id}/tests`. */
export async function listTests(diagnosisSessionId: string): Promise<DiagnosticTestRow[]> {
  const raw = await api.get<{
    tests: Array<{
      test_id: string; assay: string; status: string; discriminates_hypotheses: boolean;
      expected_information_gain: string; plans: Array<{ plan_id: string; readiness: string }>;
    }>;
  }>(`/api/diagnosis/sessions/${diagnosisSessionId}/tests`);
  return raw.tests.map((t) => ({
    testId: t.test_id, assay: t.assay, status: t.status, discriminatesHypotheses: t.discriminates_hypotheses,
    expectedInformationGain: t.expected_information_gain,
    plans: t.plans.map((p) => ({ planId: p.plan_id, readiness: p.readiness })),
  }));
}

/** `GET /api/diagnosis/sessions/{id}/decisions`. */
export async function listDecisions(diagnosisSessionId: string): Promise<DecisionRow[]> {
  const raw = await api.get<{
    decisions: Array<{
      decision_id: string; diagnosis_version: number; stopping_reason: string; allowed_next_action: string;
      handoff_status: string; leading_hypothesis_ids: string[]; alternatives_not_excluded_ids: string[];
    }>;
  }>(`/api/diagnosis/sessions/${diagnosisSessionId}/decisions`);
  return raw.decisions.map((d) => ({
    decisionId: d.decision_id, diagnosisVersion: d.diagnosis_version, stoppingReason: d.stopping_reason,
    allowedNextAction: d.allowed_next_action, handoffStatus: d.handoff_status,
    leadingHypothesisIds: d.leading_hypothesis_ids, alternativesNotExcludedIds: d.alternatives_not_excluded_ids,
  }));
}

/** `POST /api/diagnosis/decisions/{id}/approve`. */
export async function approveDecision(decisionId: string, actorId: string, approved: boolean, reason = ""): Promise<{ decisionId: string; handoffStatus: string }> {
  const r = await api.post<{ decision_id: string; handoff_status: string }>(`/api/diagnosis/decisions/${decisionId}/approve`, {
    decision_id: decisionId, actor_id: actorId, approved, reason,
  });
  return { decisionId: r.decision_id, handoffStatus: r.handoff_status };
}

/** `GET /api/diagnosis/sessions/{id}/audit-trail`. */
export async function getAuditTrail(diagnosisSessionId: string): Promise<DiagnosisTransitionRow[]> {
  const raw = await api.get<{ transitions: Array<{ state: string; selected_next_state: string | null; gate_result: Record<string, unknown> | null; started_at: number }> }>(
    `/api/diagnosis/sessions/${diagnosisSessionId}/audit-trail`,
  );
  return raw.transitions.map((t) => ({ state: t.state, selectedNextState: t.selected_next_state, gateResult: t.gate_result, startedAt: t.started_at }));
}

/** `GET /api/diagnosis/sessions/{id}/report`. */
export async function getReport(diagnosisSessionId: string): Promise<ReportSection[]> {
  const raw = await api.get<{ sections: Array<{ title: string; content: Record<string, unknown>; trace_ids: string[] }> }>(
    `/api/diagnosis/sessions/${diagnosisSessionId}/report`,
  );
  return raw.sections.map((s) => ({ title: s.title, content: s.content, traceIds: s.trace_ids }));
}

/** Pure state-transition actions - `POST /api/diagnosis/sessions/{id}/action/{action}`. See this module's
 * docstring: these never author new content, only move `status` forward. */
export const DIAGNOSIS_SESSION_ACTIONS = [
  "mark_hypotheses_generated", "mark_evidence_assessed", "mark_model_evidence_pending", "mark_hypotheses_ranked",
  "enter_model_conflicted", "enter_test_selection_required", "select_test", "enter_awaiting_test_result",
  "ingest_test_result_and_update_belief", "resolve_human_review", "reopen_diagnosis", "close_diagnosis",
] as const;
export type DiagnosisSessionAction = (typeof DIAGNOSIS_SESSION_ACTIONS)[number];

export async function sessionAction(
  diagnosisSessionId: string,
  action: DiagnosisSessionAction,
  actorId: string,
  kwargs: Record<string, unknown> = {},
): Promise<{ diagnosisSessionId: string; status: string }> {
  const r = await api.post<{ diagnosis_session_id: string; status: string }>(
    `/api/diagnosis/sessions/${diagnosisSessionId}/action/${action}`,
    { actor_id: actorId, kwargs },
  );
  return { diagnosisSessionId: r.diagnosis_session_id, status: r.status };
}

File: harness/api/diagnosis.py
"""Bottleneck Diagnosis Loop API routes (doc03 §9). Read/write against the
real workflow, not a demo page - every route calls the same service
functions the end-to-end smoke tests and integration tests exercise.
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.orm import Session

from harness.api.deps import get_db_session
from harness.diagnosis import decision_service as dec_svc
from harness.diagnosis import evidence as evidence_svc
from harness.diagnosis import model_service as model_svc
from harness.diagnosis import service as diag_svc
from harness.diagnosis.loop import DiagnosisGateRejectedError, DiagnosisLoopController, IllegalDiagnosisTransitionError
from harness.diagnosis.model_adapters.registry import detect_all_capabilities
from harness.diagnosis.models import (
    DiagnosisDecision, DiagnosisSession, DiagnosisTransition, EngineeringProblem, EvidenceItem, EvidenceLink, HypothesisAssessment,
)
from harness.diagnosis.grounding import GroundingError, derive_engineering_problem, evaluate_observation_grounding
from harness.diagnosis.report import render_report
from harness.learning.models import HypothesisVersion
from harness.memory import event_types as et
from harness.projects.models import ProjectEvent

router = APIRouter(prefix="/api/diagnosis", tags=["diagnosis"])
_loop = DiagnosisLoopController()


class CreateSessionBody(BaseModel):
    project_id: str
    actor_id: str
    workflow_run_id: str | None = None
    triggering_failure_case_id: str | None = None
    objective_id: str | None = None
    biological_system: dict[str, Any] = {}
    baseline_observation_ids: list[str] = []
    observation_ids: list[str] = []


@router.post("/sessions")
def create_session(body: CreateSessionBody, session: Session = Depends(get_db_session)) -> dict:
    data = body.model_dump()
    observation_ids = data.pop("observation_ids")
    sess = diag_svc.start_diagnosis_session(session, **data)
    sess.pending_request_context = {**(sess.pending_request_context or {}), "observation_ids": observation_ids}
    session.flush()
    return {"diagnosis_session_id": sess.diagnosis_session_id, "status": sess.status}


@router.get("/sessions/{diagnosis_session_id}")
def get_session(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    sess = diag_svc.get_session(session, diagnosis_session_id)
    if sess is None:
        raise HTTPException(404, "diagnosis session not found")
    return {
        "diagnosis_session_id": sess.diagnosis_session_id, "project_id": sess.project_id, "status": sess.status,
        "data_sufficiency": sess.data_sufficiency, "approval_state": sess.approval_state,
        "active_hypothesis_set_version": sess.active_hypothesis_set_version, "biological_system": sess.biological_system,
        "baseline_observation_ids": sess.baseline_observation_ids, "version": sess.version,
    }


class DeriveEngineeringProblemBody(BaseModel):
    observation_id: str
    comparison_observation_id: str
    abnormality_statement: str | None = None


@router.get("/sessions/{diagnosis_session_id}/grounding")
def get_grounding(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    try:
        return evaluate_observation_grounding(session, diagnosis_session_id).to_dict()
    except GroundingError as e:
        raise HTTPException(404, str(e))


@router.post("/sessions/{diagnosis_session_id}/engineering-problems/derive")
def derive_problem(diagnosis_session_id: str, body: DeriveEngineeringProblemBody, session: Session = Depends(get_db_session)) -> dict:
    try:
        p = derive_engineering_problem(session, diagnosis_session_id=diagnosis_session_id, **body.model_dump())
    except GroundingError as e:
        raise HTTPException(422, str(e))
    return {"engineering_problem_id": p.engineering_problem_id, "status": p.status}


@router.get("/sessions/{diagnosis_session_id}/engineering-problems")
def list_engineering_problems(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    rows = session.execute(select(EngineeringProblem).where(EngineeringProblem.diagnosis_session_id == diagnosis_session_id)).scalars().all()
    return {"engineering_problems": [{c.name: getattr(p, c.name) for c in p.__table__.columns} for p in rows]}


@router.get("/sessions")
def list_sessions_for_project(project_id: str, session: Session = Depends(get_db_session)) -> dict:
    rows = session.execute(
        select(DiagnosisSession).where(DiagnosisSession.project_id == project_id).order_by(DiagnosisSession.created_at.desc())
    ).scalars().all()
    return {
        "sessions": [
            {
                "diagnosis_session_id": s.diagnosis_session_id, "project_id": s.project_id, "status": s.status,
                "data_sufficiency": s.data_sufficiency, "approval_state": s.approval_state,
                "biological_system": s.biological_system, "created_at": s.created_at, "updated_at": s.updated_at,
            }
            for s in rows
        ]
    }


@router.get("/sessions/{diagnosis_session_id}/hypotheses")
def list_hypotheses(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    assessments = session.execute(
        select(HypothesisAssessment).where(HypothesisAssessment.diagnosis_session_id == diagnosis_session_id)
    ).scalars().all()
    hyp_ids = {a.hypothesis_version_id for a in assessments}
    hyps = {
        h.hypothesis_version_id: h
        for h in session.execute(select(HypothesisVersion).where(HypothesisVersion.hypothesis_version_id.in_(hyp_ids))).scalars()
    } if hyp_ids else {}
    return {
        "hypotheses": [
            {
                "hypothesis_version_id": a.hypothesis_version_id, "status": a.status,
                "mechanism_class": hyps.get(a.hypothesis_version_id).mechanism_class if a.hypothesis_version_id in hyps else None,
                "statement": hyps.get(a.hypothesis_version_id).statement if a.hypothesis_version_id in hyps else None,
                "explanatory_coverage": a.explanatory_coverage, "contradictions": a.contradictions,
                # Module 2 §7/§9: already computed by hypothesis_generator.py
                # and persisted on HypothesisVersion, but this route never
                # returned them before - falsifiers/omics_layers stayed
                # invisible even after a successful diagnosis.
                "falsifiers": hyps.get(a.hypothesis_version_id).falsifiers if a.hypothesis_version_id in hyps else [],
                "omics_layers": (hyps.get(a.hypothesis_version_id).scope or {}).get("omics_layers", []) if a.hypothesis_version_id in hyps else [],
            }
            for a in assessments
        ]
    }


def _latest_reviews(session: Session, evidence_link_ids: list[str]) -> dict[str, dict]:
    """Latest human review verdict per evidence link, read back off the
    `ProjectEvent` ledger (see `evidence_svc.review_evidence_link`'s
    docstring for why this isn't a mutable column)."""
    if not evidence_link_ids:
        return {}
    rows = session.execute(
        select(ProjectEvent)
        .where(ProjectEvent.entity_type == "EvidenceLink", ProjectEvent.entity_id.in_(evidence_link_ids), ProjectEvent.event_type == et.DIAGNOSIS_EVIDENCE_LINK_REVIEWED)
        .order_by(ProjectEvent.seq)
    ).scalars().all()
    latest: dict[str, dict] = {}
    for r in rows:
        latest[r.entity_id] = r.payload  # ordered by seq ascending, so the last write per id wins
    return latest


@router.get("/sessions/{diagnosis_session_id}/evidence")
def list_evidence(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    assessments = session.execute(
        select(HypothesisAssessment).where(HypothesisAssessment.diagnosis_session_id == diagnosis_session_id)
    ).scalars().all()
    hyp_ids = {a.hypothesis_version_id for a in assessments}
    links = session.execute(select(EvidenceLink).where(EvidenceLink.hypothesis_version_id.in_(hyp_ids))).scalars().all() if hyp_ids else []
    reviews = _latest_reviews(session, [l.evidence_link_id for l in links])
    return {
        "evidence_links": [
            {
                "evidence_link_id": l.evidence_link_id, "hypothesis_version_id": l.hypothesis_version_id,
                "evidence_item_id": l.evidence_item_id, "relation": l.relation, "claim": l.claim,
                "review_status": reviews[l.evidence_link_id]["verdict"] if l.evidence_link_id in reviews else "unreviewed",
                "review_note": reviews[l.evidence_link_id].get("note", "") if l.evidence_link_id in reviews else "",
            }
            for l in links
        ]
    }


class ReviewEvidenceLinkBody(BaseModel):
    verdict: str
    actor_id: str
    note: str = ""


@router.post("/evidence-links/{evidence_link_id}/review")
def review_evidence_link(evidence_link_id: str, body: ReviewEvidenceLinkBody, session: Session = Depends(get_db_session)) -> dict:
    try:
        evidence_svc.review_evidence_link(session, evidence_link_id=evidence_link_id, **body.model_dump())
    except evidence_svc.InvalidReviewVerdict as e:
        raise HTTPException(422, str(e))
    except ValueError as e:
        raise HTTPException(404, str(e))
    return {"evidence_link_id": evidence_link_id, "verdict": body.verdict}


class LinkEvidenceBody(BaseModel):
    hypothesis_version_id: str
    evidence_item_id: str
    relation: str
    actor_id: str
    claim: str = ""
    condition_match: str = "unknown"


@router.post("/evidence-links")
def link_evidence(body: LinkEvidenceBody, session: Session = Depends(get_db_session)) -> dict:
    try:
        link = evidence_svc.link_evidence(session, **body.model_dump())
    except evidence_svc.InvalidEvidenceRelation as e:
        raise HTTPException(422, str(e))
    return {"evidence_link_id": link.evidence_link_id}


@router.get("/evidence-items")
def list_evidence_items(project_id: str, session: Session = Depends(get_db_session)) -> dict:
    """So the frontend can offer a picker instead of asking the user to
    remember/type an `evidence_item_id` by hand."""
    rows = session.execute(
        select(EvidenceItem).where(EvidenceItem.project_id == project_id).order_by(EvidenceItem.created_at.desc())
    ).scalars().all()
    return {
        "evidence_items": [
            {
                "evidence_item_id": r.evidence_item_id, "source_type": r.source_type,
                "source_reference": r.source_reference, "content_summary": r.content_summary,
                "quality": r.quality, "directness": r.directness, "created_at": r.created_at,
            }
            for r in rows
        ]
    }


class CreateEvidenceItemBody(BaseModel):
    project_id: str
    actor_id: str
    source_type: str
    content_summary: str
    source_reference: str | None = None
    quality: str = "low"
    directness: str = "indirect"


@router.post("/evidence-items")
def create_evidence_item(body: CreateEvidenceItemBody, session: Session = Depends(get_db_session)) -> dict:
    """`evidence_item_id` is always server-generated (`evidence_svc.record_evidence_item`
    -> `new_id("EVID")`) - the request body never carries a caller-supplied id."""
    item = evidence_svc.record_evidence_item(session, **body.model_dump())
    return {"evidence_item_id": item.evidence_item_id}


@router.get("/model-capabilities")
def model_capabilities() -> dict:
    caps = detect_all_capabilities()
    return {name: {"available": c.available, "reason": c.reason} for name, c in caps.items()}


class RunModelBody(BaseModel):
    project_id: str
    diagnosis_session_id: str
    adapter_name: str
    actor_id: str
    inputs: dict[str, Any] = {}
    context: dict[str, Any] = {}
    constraints_objective_parameters: dict[str, Any] = {}
    sensitivity_variant_of: str | None = None


@router.post("/model-runs")
def run_model(body: RunModelBody, session: Session = Depends(get_db_session)) -> dict:
    try:
        record = model_svc.execute_model_run(session, **body.model_dump())
    except KeyError as e:
        raise HTTPException(400, str(e))
    return {
        "model_run_id": record.model_run_id, "capability_status": record.capability_status,
        "runtime_status": record.runtime_status, "outputs": record.outputs, "log_summary": record.log_summary,
    }


@router.get("/sessions/{diagnosis_session_id}/tests")
def list_tests(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    from harness.diagnosis.models import DiagnosticTest, ExperimentalExecutionPlan

    tests = session.execute(select(DiagnosticTest).where(DiagnosticTest.diagnosis_session_id == diagnosis_session_id)).scalars().all()
    result = []
    for t in tests:
        plans = session.execute(select(ExperimentalExecutionPlan).where(ExperimentalExecutionPlan.diagnostic_test_id == t.test_id)).scalars().all()
        result.append({
            "test_id": t.test_id, "assay": t.assay, "status": t.status, "discriminates_hypotheses": t.discriminates_hypotheses,
            "expected_information_gain": t.expected_information_gain,
            "plans": [{"plan_id": p.plan_id, "readiness": p.readiness} for p in plans],
        })
    return {"tests": result}


class HumanApprovalBody(BaseModel):
    decision_id: str
    actor_id: str
    approved: bool
    reason: str = ""


@router.post("/decisions/{decision_id}/approve")
def approve_decision(decision_id: str, body: HumanApprovalBody, session: Session = Depends(get_db_session)) -> dict:
    status = "approved" if body.approved else "rejected"
    try:
        d = dec_svc.set_handoff_status(
            session, decision_id=decision_id, handoff_status=status, actor_id=body.actor_id,
            human_approval={"approver": body.actor_id, "decision": status, "reason": body.reason},
        )
    except ValueError as e:
        raise HTTPException(404, str(e))
    return {"decision_id": d.decision_id, "handoff_status": d.handoff_status}


@router.get("/sessions/{diagnosis_session_id}/decisions")
def list_decisions(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    rows = session.execute(
        select(DiagnosisDecision).where(DiagnosisDecision.diagnosis_session_id == diagnosis_session_id).order_by(DiagnosisDecision.diagnosis_version)
    ).scalars().all()
    return {
        "decisions": [
            {
                "decision_id": d.decision_id, "diagnosis_version": d.diagnosis_version, "stopping_reason": d.stopping_reason,
                "allowed_next_action": d.allowed_next_action, "handoff_status": d.handoff_status,
                "leading_hypothesis_ids": d.leading_hypothesis_ids, "alternatives_not_excluded_ids": d.alternatives_not_excluded_ids,
            }
            for d in rows
        ]
    }


@router.get("/sessions/{diagnosis_session_id}/audit-trail")
def audit_trail(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    rows = session.execute(
        select(DiagnosisTransition).where(DiagnosisTransition.diagnosis_session_id == diagnosis_session_id).order_by(DiagnosisTransition.started_at)
    ).scalars().all()
    return {
        "transitions": [
            {"state": t.state, "selected_next_state": t.selected_next_state, "gate_result": t.gate_result, "started_at": t.started_at}
            for t in rows
        ]
    }


@router.get("/sessions/{diagnosis_session_id}/report")
def get_report(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    try:
        report = render_report(session, diagnosis_session_id=diagnosis_session_id)
    except ValueError as e:
        raise HTTPException(404, str(e))
    return report.to_dict()


@router.get("/sessions/{diagnosis_session_id}/mechanism-graph")
def get_mechanism_graph(diagnosis_session_id: str, session: Session = Depends(get_db_session)) -> dict:
    """Module 2 (Engineering Decision Intelligence Layer prompt §8): the
    Engineering Reasoning Graph, read-only. `build_mechanism_graph`'s output
    is not persisted anywhere (`HypothesisVersion.mechanism_graph_ref` is a
    bare reference string no writer ever populates) - this recomputes the
    same, real, deterministic graph the diagnosis pipeline itself built from
    the project's own `target_product`/`host_definition`, never a fabricated
    or cached-stale view.
    """
    from dataclasses import asdict

    from harness.diagnosis.mechanism_graph import build_mechanism_graph
    from harness.projects.models import Project

    sess = diag_svc.get_session(session, diagnosis_session_id)
    if sess is None:
        raise HTTPException(404, "diagnosis session not found")
    proj = session.get(Project, sess.project_id)
    host = " ".join(str(v) for v in (proj.host_definition or {}).values() if v) if proj else "unknown"
    product = proj.target_product if proj else "unknown"
    graph = build_mechanism_graph(
        phenotype=f"bottleneck diagnosis for {product}", product=product or "unknown", host=host or "unknown",
    )
    return {
        "nodes": [asdict(n) for n in graph.nodes],
        "edges": [asdict(e) for e in graph.edges],
        "unknowns": graph.unknowns,
    }


_SESSION_ACTIONS = (
    "mark_hypotheses_generated", "mark_evidence_assessed", "mark_model_evidence_pending", "mark_hypotheses_ranked",
    "enter_model_conflicted", "enter_test_selection_required", "select_test", "enter_awaiting_test_result",
    "ingest_test_result_and_update_belief", "resolve_human_review", "reopen_diagnosis", "close_diagnosis",
)


class SessionActionBody(BaseModel):
    actor_id: str
    kwargs: dict[str, Any] = {}


@router.post("/sessions/{diagnosis_session_id}/action/{action}")
def session_action(diagnosis_session_id: str, action: str, body: SessionActionBody, session: Session = Depends(get_db_session)) -> dict:
    if action not in _SESSION_ACTIONS:
        raise HTTPException(400, f"unknown or unsupported action {action!r}; must be one of {_SESSION_ACTIONS}")
    sess = diag_svc.get_session(session, diagnosis_session_id)
    if sess is None:
        raise HTTPException(404, "diagnosis session not found")
    fn = getattr(_loop, action)
    try:
        fn(session, sess, actor_id=body.actor_id, **body.kwargs)
    except IllegalDiagnosisTransitionError as e:
        raise HTTPException(409, str(e))
    except DiagnosisGateRejectedError as e:
        raise HTTPException(422, str(e))
    return {"diagnosis_session_id": sess.diagnosis_session_id, "status": sess.status}
